src/Development/IDE/Core/Tracing.hs

```diff
@@ -0,0 +1,77 @@
+module Development.IDE.Core.Tracing
+    ( otTraced
+    , otTracedHandler
+    , otTracedAction
+    , startTelemetry
+    )
+where
+
+import           Control.Concurrent
+import           Control.Exception
+import           Control.Monad
+import           Development.Shake
+import           Development.IDE.Types.Options
+import           Language.Haskell.LSP.Types
+import           OpenTelemetry.Eventlog
+import qualified Data.ByteString.Char8         as BS
+import           Control.Concurrent.Extra
+import           Control.Concurrent.Async
+import           HeapSize
+import qualified Data.HashMap.Strict as HMap
+import           Data.Functor
+import           System.Mem (performGC)
+
+-- | Trace an action using OpenTelemetry. Adds various useful info into tags in the OpenTelemetry span.
+otTraced :: BS.ByteString -> IO a -> IO a
+otTraced name act = bracket (beginSpan name) endSpan (const act)
+
+-- | Trace a handler using OpenTelemetry. Adds various useful info into tags in the OpenTelemetry span.
+otTracedHandler
+    :: String -- ^ Message type
+    -> String -- ^ Message label
+    -> IO a
+    -> IO a
+otTracedHandler requestType label act =
+  let !name =
+        if null label
+          then BS.pack requestType
+          else BS.pack (requestType <> ":" <> show label)
+   -- Add an event so all requests can be quickly seen in the viewer without searching
+   in withSpan name (\sp -> addEvent sp "" (name <> " received") >> act)
+
+-- | Trace a Shake action using opentelemetry.
+otTracedAction
+    :: Show k
+    => k -- ^ The Action's Key
+    -> NormalizedFilePath -- ^ Path to the file the action was run for
+    -> Action a -- ^ The action
+    -> Action a
+otTracedAction key file act = actionBracket
+    (do
+        sp <- beginSpan (BS.pack (show key))
+        setTag sp "File" (BS.pack $ fromNormalizedFilePath file)
+        return sp
+    )
+    endSpan
+    (const act)
+
+startTelemetry :: Show k => String -> Var (HMap.HashMap k v) -> IdeOTProfiling -> IO ()
+startTelemetry _ _ (IdeOTProfiling False) = return ()
+startTelemetry name valuesRef (IdeOTProfiling True) = do
+  mapBytesInstrument <- mkValueObserver (BS.pack name <> " size_bytes")
+  mapCountInstrument <- mkValueObserver (BS.pack name <> " count")
```

Comment:
In addition to these global counters, it would be very useful to have per-rule counters, in order to show which rules are expensive (in items and in memory)

Reply:
I agree with all the suggested changes, with one caveat: I don't think it's feasible to measure per-rule allocations with the current method, as it would be too slow, even for benchmarking use.

Reply:
Would a measure of per-rule allocations and deriving the total from that be fast enough?

Reply:
I'm going to try implementing this and see the results. My guess is it would actually be quite a bit faster because now you are counting items only when they are inserted into the map. However now the metric becomes one of total allocations instead of live memory since you miss deletions. Would there be some way "intercepting" those so you can count the size of deleted elements?

Reply:
Hmm, I don't think it's going to be easy (or possible) to observe the deletions. 

Maybe what we want is to keep the current behaviour, but instead of getting the recursive size of the whole values map, break it down by key type and get the recursive size of each group

Reply:
So, as a test, I tried measuring allocations when objects get added to the `state` map, (https://github.com/mpardalos/ghcide/compare/opentelemetry-memory...detailed-memory-test), and it turns out I was wrong. Running a small "hover" benchmark (see code), took about 3 hours, which I think qualifies as unusable for now. 

> Maybe what we want is to keep the current behaviour, but instead of getting the recursive size of the whole values map, break it down by key type and get the recursive size of each group

That is probably much more feasible. I will give that a try as well and report back

Reply:
Why is it so slow?

Reply:
I tried analyzing which entries in the map took the longest and it seems like the worst offenders are GhcSessionIO and GhcSession entries, each of which can take about 10s to measure, I think because their representation is very nested (each can be >1000000 closures). If there is a way to estimate the size of these entries instead of having to traverse them it could improve performance a lot. 

If there isn't a way to handle this then the only way to improve performance would be to optimise heapsize further.

Reply:
Have you tried the group approach below? I think that's probably what we want to measure and not the individual insertions, since I suspect that when inserting a new item in the map part of its cost is shared with other entries in the map. 

> Maybe what we want is to keep the current behaviour, but instead of getting the recursive size of the whole values map, break it down by key type and get the recursive size of each group